### PR TITLE
[public-api] Update getWorkspace api response implementation

### PR DESCRIPTION
**Pull request by jeanp413**
## Description

Update getWorkspace api response with new `recent_folders` attribute

#### Preview status

gitpod:summary

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [x] with-ws-manager-mk2
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [ ] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`
</details>
 
/hold

### CODE REVIEW

File: components/public-api-server/pkg/apiv1/workspace.go [hunk 1]
```diff
@@ -403,6 +444,21 @@ func convertWorkspaceInstance(wsi *protocol.WorkspaceInstance, shareable bool) (
...
+		location = config.WorkspaceLocation
+		if location == "" {
+			location = config.CheckoutLocation
+		}
+	}
+	if location == "" && wsCtx != nil && wsCtx.Repository != nil {
+		location = wsCtx.Repository.Name
+
+	}
+	recentFolders = append(recentFolders, filepath.Join("/workspace", location))
```
Comment (jeanp413) [inline components/public-api-server/pkg/apiv1/workspace.go:460]: From what I understood reading the code, the server doesn't know about `/workspace` base folder, instead it's harcoded inside ws-manager, is it fine to harcode it here too?

File: components/public-api-server/pkg/apiv1/workspace.go [hunk 2]
```diff
@@ -46,7 +48,13 @@ func (s *WorkspaceService) GetWorkspace(ctx context.Context, req *connect.Reques
 		return nil, proxy.ConvertError(err)
 	}
 
-	workspace, err := convertWorkspaceInfo(ws)
+	authProviders, err := conn.GetAuthProviders(ctx)
+	if err != nil {
+		log.Extract(ctx).WithError(err).Error("Failed to get auth providers.")
+		return nil, proxy.ConvertError(err)
+	}
```
Comment (jeanp413) [inline components/public-api-server/pkg/apiv1/workspace.go:55]: I'm not sure if adding another api call here is fine

File: components/public-api-server/pkg/apiv1/workspace.go [hunk 3]
```diff
@@ -229,7 +243,13 @@ func (s *WorkspaceService) StartWorkspace(ctx context.Context, req *connect.Requ
 		return nil, proxy.ConvertError(err)
 	}
 
-	workspace, err := convertWorkspaceInfo(ws)
+	authProviders, err := conn.GetAuthProviders(ctx)
```
Comment (easyCZ) [inline components/public-api-server/pkg/apiv1/workspace.go:246]: What stops us from returning this is part of `conn.GetWorkspace` from `server`?

File: components/public-api-server/pkg/apiv1/workspace.go (hunk 1 above)
Comment (easyCZ) [reply components/public-api-server/pkg/apiv1/workspace.go:460]: What's the ultimate source of truth for this information? If it's ws-manager, we should probably update ws-manager to report that directory also such that there's a single source of truth

Comment (jeanp413) [reply components/public-api-server/pkg/apiv1/workspace.go:460]: it's harcoded in many places, unsure which one is the source of truth, maybe workspace team can point it out cc @kylos101

File: components/public-api-server/pkg/apiv1/workspace.go (hunk 3 above)
Comment (jeanp413) [reply components/public-api-server/pkg/apiv1/workspace.go:246]: @easyCZ Not much but types from protocol are also used inside workspace db implementation so I'll need to split them, db types and protocol types, does that sounds good?

File: components/public-api-server/pkg/apiv1/workspace.go (hunk 1 above)
Comment (kylos101) [reply components/public-api-server/pkg/apiv1/workspace.go:460]: @jeanp413 I would consider the fact that this is hard coded to be non-blocking for this PR, as this is hard coded in webapp, workspace, and IDE components.

[webapp](https://github.com/gitpod-io/gitpod/blob/badbe674482c864b946eda4d672d9efd96fced27/components/server/src/workspace/config-provider.ts#L38)

[IDE](https://github.com/gitpod-io/gitpod/blob/badbe674482c864b946eda4d672d9efd96fced27/components/supervisor/pkg/terminal/service.go#L36)

[workspace](https://github.com/gitpod-io/gitpod/blob/badbe674482c864b946eda4d672d9efd96fced27/components/ws-manager-mk2/controllers/create.go#L39)

**Comment by jeanp413** [comment]
@easyCZ Updated the PR to implement only `RecentFolders` for now to unblock myself, I'll continue working on the git provider and adding other properties in follow up PR

File: components/public-api-server/pkg/apiv1/workspace.go [hunk 4]
```diff
@@ -403,6 +405,21 @@ func convertWorkspaceInstance(wsi *protocol.WorkspaceInstance, shareable bool) (
...
+		location = config.WorkspaceLocation
+		if location == "" {
+			location = config.CheckoutLocation
+		}
+	}
+	if location == "" && wsCtx != nil && wsCtx.Repository != nil {
+		location = wsCtx.Repository.Name
+
+	}
+	recentFolders = append(recentFolders, filepath.Join("/workspace", location))
```
Comment (easyCZ) [inline components/public-api-server/pkg/apiv1/workspace.go:421]: Looks like tests do not cover scenario for where `config.WorkspaceLocation` or `config.CheckoutLocation` is populated

File: components/gitpod-protocol/go/gitpod-service.go [hunk 5]
```diff
@@ -1730,6 +1730,10 @@ type WorkspaceContext struct {
 	ForceCreateNewWorkspace bool   `json:"forceCreateNewWorkspace,omitempty"`
 	NormalizedContextURL    string `json:"normalizedContextURL,omitempty"`
 	Title                   string `json:"title,omitempty"`
+
+	// Commit context
+	Repository *Repository `json:"repository,omitempty"`
```
Comment (easyCZ) [inline components/gitpod-protocol/go/gitpod-service.go:1735]: When is it possible for a workspace to not have a Repository?

Comment (jeanp413) [reply components/gitpod-protocol/go/gitpod-service.go:1735]: I'm just following the [protocol definition](https://github.com/gitpod-io/gitpod/blob/01b23af10e7241abc588b1c23bda31bf9b2a9f7c/components/gitpod-protocol/src/protocol.ts), so `WorkspaceContext` interface does not have `Repository` but more specific contexts like `CommitContext` extends from it, I could define all other interfaces but then I think I will need to define a custom json parser so I can check for specific properties and return the correct struct type.
Also If it's expected for `getWorkspace` to always return a context of type `CommitContext` then shouldn't the typing in protocol.ts be updated? what about snapshot context?

File: components/public-api-server/pkg/apiv1/workspace.go (hunk 4 above)
Comment (jeanp413) [reply components/public-api-server/pkg/apiv1/workspace.go:421]: done

**Pull request merged**